Mathlib/Probability/CondVar.lean

```diff
@@ -0,0 +1,187 @@
+conditional expectation).
+-/
+
+open MeasureTheory Filter
+open scoped ENNReal
+
+namespace ProbabilityTheory
+variable {Ω : Type*} {m₀ m : MeasurableSpace Ω} {hm : m ≤ m₀} {X Y : Ω → ℝ} {μ : Measure[m₀] Ω}
+  {s : Set Ω}
+
+variable (m X μ) in
+/-- Conditional variance of a real-valued random variable. It is defined as 0 if any one of the
+following conditions is true:
+- `m` is not a sub-σ-algebra of `m₀`,
+- `μ` is not σ-finite with respect to `m`,
+- `X - μ[X | m]` is not square-integrable. -/
+noncomputable def condVar : Ω → ℝ := μ[(X - μ[X | m]) ^ 2 | m]
+
+@[inherit_doc] scoped notation "Var[" X " ; " μ " | " m "]" => condVar m X μ
+
+lemma condVar_of_not_le (hm : ¬m ≤ m₀) : Var[X ; μ | m] = 0 := by rw [condVar, condExp_of_not_le hm]
+
+lemma condVar_of_not_sigmaFinite (hμm : ¬SigmaFinite (μ.trim hm)) :
+    Var[X ; μ | m] = 0 := by rw [condVar, condExp_of_not_sigmaFinite hm hμm]
+
+open scoped Classical in
+lemma condVar_of_sigmaFinite [SigmaFinite (μ.trim hm)] :
+    Var[X ; μ | m] =
+      if Integrable (fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2) μ then
+        if StronglyMeasurable[m] (fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2) then
+          fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2
+        else aestronglyMeasurable_condExpL1.mk (condExpL1 hm μ fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2)
+      else 0 := condExp_of_sigmaFinite _
+
+lemma condVar_of_stronglyMeasurable [SigmaFinite (μ.trim hm)]
+    (hX : StronglyMeasurable[m] X) (hXint : Integrable ((X - μ[X | m]) ^ 2) μ) :
+    Var[X ; μ | m] = fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2 :=
+  condExp_of_stronglyMeasurable _ ((hX.sub stronglyMeasurable_condExp).pow _) hXint
+
+lemma condVar_of_not_integrable (hXint : ¬ Integrable (fun ω ↦ (X ω - (μ[X | m]) ω) ^ 2) μ) :
+    Var[X ; μ | m] = 0 := condExp_of_not_integrable hXint
+
+@[simp]
+lemma condVar_zero : Var[0 ; μ | m] = 0 := by
+  by_cases hm : m ≤ m₀
+  swap; · rw [condVar_of_not_le hm]
+  by_cases hμm : SigmaFinite (μ.trim hm)
+  swap; · rw [condVar_of_not_sigmaFinite hμm]
+  rw [condVar_of_stronglyMeasurable stronglyMeasurable_zero]
+  · simp [← Pi.zero_def]
+  · simpa [-integrable_zero] using integrable_zero ..
+
+@[simp]
+lemma condVar_const (hm : m ≤ m₀) (c : ℝ) : Var[fun _ ↦ c ; μ | m] = 0 := by
+  obtain rfl | hc := eq_or_ne c 0
+  · simp [← Pi.zero_def]
+  by_cases hμm : IsFiniteMeasure μ
+  · simp [condVar, hm, Pi.pow_def]
+  · simp [condVar, condExp_of_not_integrable, integrable_const_iff_isFiniteMeasure hc,
+      integrable_const_iff_isFiniteMeasure <| pow_ne_zero _ hc, hμm, Pi.pow_def]
+
+lemma condVar_ae_eq_condExpL1 [hμm : SigmaFinite (μ.trim hm)] (X : Ω → ℝ) :
+    Var[X ; μ | m] =ᵐ[μ] condExpL1 hm μ ((X - μ[X | m]) ^ 2) :=
+  condExp_ae_eq_condExpL1 ..
+
+lemma condVar_ae_eq_condExpL1CLM [SigmaFinite (μ.trim hm)]
+    (hX : Integrable ((X - μ[X | m]) ^ 2) μ) :
+    Var[X ; μ | m] =ᵐ[μ] condExpL1CLM _ hm μ (hX.toL1 <| (X - μ[X | m]) ^ 2) :=
+  condExp_ae_eq_condExpL1CLM ..
```

Comment:
Do you really need those? Ideally we would interact with `condExp` as exclusively as possible, and relegate `condExpL2`, `condExpL1` and `condExpL1CLM` to the role of intermediate objects in the construction of `condExp`. I know that sometimes they are useful and do things that `condExp` can't, but let's invoque them only when necessary.

Reply:
No, I was just copy-pasting the `condExp` API

Reply:
Sorry, my comment was not clear: I was suggesting you delete those two results, on the basis that we don't want to use the condExp variants outside of the condExp constructions, unless necessary.

Reply:
No it was clear. I have deleted them

Reply:
You mean that you deleted them in a commit that you have not pushed yet? I still see the lemmas in the PR.

Reply:
Hmm okay, apparently they were not deleted